federation/pkg/federation-controller/replicaset/planner/planer.go

```diff
@@ -0,0 +1,144 @@
+		return &namedLocalReplicaSetPreferences{
+			clusterName:                name,
+			LocalReplicaSetPreferences: pref,
+		}
+	}
+
+	for _, cluster := range availableClusters {
+		if localRSP, found := p.preferences.Clusters[cluster]; found {
+			preferences = append(preferences, named(cluster, localRSP))
+		} else {
+			if localRSP, found := p.preferences.Clusters["*"]; found {
+				preferences = append(preferences, named(cluster, localRSP))
+			} else {
+				preferences = append(preferences, named(cluster, fed_api.LocalReplicaSetPreferences{Weight: 0}))
+			}
+		}
+	}
+	sort.Sort(byWeight(preferences))
+
+	remainingReplicas := replicasToDistribute
+	plan := make(map[string]int64, len(preferences))
+
+	// Assign each cluster the minimum number replica it requested.
+	for _, preference := range preferences {
+		min := minInt64(preference.MinReplicas, remainingReplicas)
+		remainingReplicas -= min
+		plan[preference.clusterName] = min
+	}
+
+	modified := true
+
+	// It is possible single pass of the loop is not enough to distribue all replicas among clusters due
+	// to weight, max and rounding corner cases. In such case we iterate until either
+	// there is no replicas or no cluster gets any more replicas or the number
+	// of attempts is less than available couster count. Every loop either distributes all remainingReplicas
+	// or maxes out at least one cluster.
+	// TODO: This algorithm is O(clusterCount^2). Find something faster.
+	for trial := 0; trial < len(availableClusters) && modified && remainingReplicas > 0; trial++ {
+		modified = false
+		weightSum := int64(0)
+		for _, preference := range preferences {
+			weightSum += preference.Weight
+		}
+		newPreferences := make([]*namedLocalReplicaSetPreferences, 0, len(preferences))
+
+		distributeInThisLoop := remainingReplicas
+		for _, preference := range preferences {
+			if weightSum > 0 {
+				start := plan[preference.clusterName]
+				// Distribute the remaining replicas, rounding fractions up from .5 and down from .5
+				extra := (distributeInThisLoop*preference.Weight + weightSum/int64(2)) / weightSum
+				extra = minInt64(extra, remainingReplicas)
+				// In total there should be the amount that was there at start plus whatever is due
+				// in this iteration
+				total := start + extra
+
+				// Check if we don't overflow the cluster, and if yes don't consider this cluster
+				// in any of the following iterations.
+				if preference.MaxReplicas != nil && total > *preference.MaxReplicas {
+					total = *preference.MaxReplicas
+				} else {
+					newPreferences = append(newPreferences, preference)
+				}
+
+				// Only total-start replicas were actually taken.
+				remainingReplicas -= (total - start)
+				plan[preference.clusterName] = total
+
+				// Something extra got scheduled on this cluster.
+				if total > start {
+					modified = true
+				}
+			} else {
+				break
+			}
+		}
+		preferences = newPreferences
+	}
+
+	return plan
+}
+
+func minInt64(a int64, b int64) int64 {
```

Comment:
It seems very sad that you have to write this, but I guess there is no other way in golang.  This makes me sad :-(